core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java

```diff
@@ -386,4 +389,21 @@ public void testThatUnrecognizedFromStringThrowsException() {
             assertThat(e.getMessage(), is("No score mode for child query [unrecognized value] found"));
         }
     }
+
+    public void testIgnoreUnmapped() throws IOException {
+        HasChildQueryBuilder queryBuilder = new HasChildQueryBuilder("unmapped", new MatchAllQueryBuilder());
+        queryBuilder.ignoreUnmapped(true);
+        Query query = queryBuilder.toQuery(queryShardContext());
+        assertThat(query, notNullValue());
+        assertThat(query, instanceOf(MatchNoDocsQuery.class));
+
+        queryBuilder = new HasChildQueryBuilder("unmapped", new MatchAllQueryBuilder());
+        queryBuilder.ignoreUnmapped(false);
+        try {
+            queryBuilder.toQuery(queryShardContext());
+            fail("expected has_children query on unmapped field to fail when ignore_unmapped is false");
+        } catch (QueryShardException e) {
+            assertThat(e.getMessage(), containsString("[" + HasChildQueryBuilder.NAME + "] no mapping found for type [unmapped]"));
+        }
```

Comment:
Can you use `expectThrows` instead? I think we should use it rather than this try:fail/catch:check_msg pattern for new code.